More thread safety annotation coverage

ajtowns opened the pull request:
In a few cases we need to use `std::mutex` rather than the sync.h primitives. But `std::lock_guard<std::mutex>` doesn't include the clang thread safety annotations unless you also use clang's C library, which means you can't indicate when variables should be guarded by `std::mutex` mutexes.

This adds an annotated version of `std::lock_guard<std::mutex>` to threadsafety.h to fix that, and modifies places where `std::mutex` is used to take advantage of the annotations.

It's based on top of #16112, and turns the thread safety comments included there into annotations.

It also changes the RAII classes in wallet/wallet.h and rpc/blockchain.cpp to just use the atomic<bool> flag for synchronisation rather than having a mutex that doesn't actually guard anything as well.

ajtowns committed:
rpc/blockchain.cpp: thread safety annotations for latestblock

ajtowns committed:
net: fMsgProcWake use LOCK instead of lock_guard

ajtowns committed:
wallet/wallet.h: Remove mutexScanning which was only protecting a single atomic bool

practicalswift commented:
Strong concept ACK: thanks for fixing this!

promag commented:
> In a few cases we need to use `std::mutex` rather than the sync.h primitives.

Can you explain why?

sipa commented:
@promag Our "default" mutexes are recursive mutexes (which you can enter multiple times from the same thread). They're easier to work with as you don't need to worry about whether to call a lock-grabbing function from either inside or outside the lock. However, recursive mutexes can't be used for condition variable waiting, so for those, we use std::mutex (which is not recursive) directly.

Longer term we should try to get rid of recursive mutexes entirely; code is much cleaner without them.

MarcoFalke commented:
How is this different from using `Mutex` and `Lock`?

E.g.

```cpp
Mutex m;
LOCK(m);

promag commented:
@sipa thanks, I was aware of that. I wasn't sure _why_ "In a few cases we need to use `std::mutex`".

ajtowns commented:
> How is this different from using `Mutex` and `Lock`?

If you do `Mutex cs;` instead of `std::mutex cs;` then you can do `LOCK(cs);` or `WAIT_LOCK(cs, g);` instead of `MutexGuard g(cs);`. In that case the type of `g` changes from a trivial subclass of `std::lock_guard` to `UniqueLock` which is a subclass of `std::unique_lock` that supports `DEBUG_LOCKCONTENTION` in which case it uses `LogPrintf` to report potential deadlocks or lock contention. Even without the extra features we have, `std::unique_lock` is potentially higher overhead than `std::lock_guard`, but `std::lock_guard` isn't unlockable except on destruction, so `unique_lock` is needed for use with condition variables.

 * We can't use `Mutex` for locking for the logging API, because that is a circular dependency.
 * We use `std::mutex` in code that implements our specialised locking behaviour got `Mutex` etc.
 * We don't use `Mutex` in `support/lockedpool.h`, probably because at present it is independent of bitcoin interfaces in general (this PR adds a dependency on threadsafety.h, so maybe just using `Mutex` would make sense at that point).
 * We use `std::mutex` instead of `Mutex` for protecting `dir_locks` in `util/system.h`, not sure if there's a circular dependency there to justify it though?
 * We also use `std::mutex` in some unit tests, which I don't think matters much.

There's some cases where we currently use `std::lock_guard<std::mutex> g(cs)` where `cs` is a `Mutex`, rather than calling `LOCK(cs);` or `WAIT_LOCK(cs, g);` That saves a bit of overhead, while losing our contention/deadlock debugging info, which I'm not sure makes much sense. I've updated this PR to change those instances to just use `LOCK(cs)` instead of `lock_guard`.

I think that means it makes sense to:

 * Use `Mutex` and `LOCK` etc as the "normal" locking mechanism.
 * Use `RecursiveMutex` (aka `CCriticalSection`) if recursive locks are needed, or it's old code that we haven't worked through yet.
 * Use `std::mutex` and `{ MutexGuard guard(mutex); ... }` for code that's used to implement `Mutex`

I've bumped this PR to replace a couple of the `lock_guard` instances that referred to our Mutex/CCriticalSection classes with `LOCK` instead of `MutexGuard`.

ryanofsky commented:
> We can't use Mutex for locking for the logging API, because that is a circular dependency.

I don't think this is true, but maybe you can explain more. I think ideally, we would just use `Mutex` and `RecursiveMutex` everywhere, since they fully support LOCK macros, lock assertions, lock annotations, deadlock detection, and so on. What exactly is the need for std::mutex and MutexGuard?

ajtowns commented:
> > We can't use Mutex for locking for the logging API, because that is a circular dependency.
> 
> I don't think this is true (that we can't allow a circular dependency), but maybe you can explain more. 

Maybe; `LOCK(m);` becomes a call to `UniqueLock` contstructor, which calls `Enter()` (or `TryEnter()`), which calls `EnterCritical()`, which calls `push_lock()` which may call `potential_deadlock_detected()` which then calls `LogPrintf`, which calls `LogPrintStr` which, when logging to a file, will then lock `m_file_mutex`. But I don't think that actually causes a problem, because I don't think any lock is ever taken while the m_file_mutex is held, so that shouldn't trigger potential_deadlock_detected.

There's also the case if you compile with `-DDEBUG_LOCKCONTENTION`. If two processes try to log simultaneously, the first one will lock `m_file_mutex` successfully, while the second one will fail at the `try_lock` phase of `UniqueLock::Enter`, which will then call `PrintLockContention` and hence `LogPrintf`, and recursively fail for the same reason until `m_file_mutex` is released, at which point things should clean up.

> I think ideally, we would just use `Mutex` and `RecursiveMutex` everywhere, since they fully support LOCK macros, lock assertions, lock annotations, deadlock detection, and so on. What exactly is the need for std::mutex and MutexGuard?

If we can switch all the uses of std::mutex to Mutex, I don't think there's a need for MutexGuard.

ajtowns committed:
rpc/blockchain.cpp: Remove g_utxosetscan mutex that is only protecting a single atomic variable

ajtowns committed:
test/checkqueue_tests: thread safety annotations

ajtowns committed:
util/system.cpp: add thread safety annotations for dir_locks

ajtowns committed:
logging: thread safety annotations

Adds LockGuard helper in threadsafety.h to replace lock_guard<mutex>
when LOCK(Mutex) isn't available for use.

ajtowns commented:
Rebased; updated to only use `MutexGuard` for logging.h.

Remaining uses of `std::mutex` and `lock_guard` are sync.h (duh), logging.h (which is used by sync.h), support/lockedpool.h (could be changed to Mutex, but I'm not confident the overhead wouldn't cause problems), tests/checkqueue_tests.cpp (complicated, and just a test) and leveldb.

sipa commented on src/test/checkqueue_tests.cpp:
```diff
@@ -290,11 +291,14 @@ BOOST_AUTO_TEST_CASE(test_CheckQueue_UniqueCheck)
+        LOCK(UniqueCheck::m);
+        bool r = true;
+        BOOST_REQUIRE_EQUAL(UniqueCheck::results.size(), COUNT);
+        for (size_t i = 0; i < COUNT; ++i)
```
Nit: if you're reflowing this code anyway, can you put { } around the then branch?

ajtowns replied:
Done

sipa commented on src/logging.h:
```diff
@@ -62,9 +63,18 @@ namespace BCLog {
-        std::list<std::string> m_msgs_before_open; // GUARDED_BY(m_cs)
-        bool m_buffering{true};                    //!< Buffer messages before logging can be started. GUARDED_BY(m_cs)
+
+        // so MutexGuard provides an annotated version of lock_guard for us
```
Maybe move MutexGuard to sync.h, as it's more generally useful?

ajtowns replied:
There are a bunch of places that include logging.h but not sync.h, so doing it via threadsafety.h might keep things a little more minimal? I've done this.

ajtowns commented:
Rebased to master to avoid #18111 and updated with @sipa's suggestions.

ajtowns commented:
Rebased for the ToString changes

ajtowns commented:
Rebased

hebasto commented:
Concept ACK.

hebasto commented on src/rpc/blockchain.cpp:
```diff
@@ -1989,18 +1988,15 @@ class CoinsViewScanReserver
     bool reserve() {
         CHECK_NONFATAL(!m_could_reserve);
-        std::lock_guard<std::mutex> lock(g_utxosetscan);
-        if (g_scan_in_progress) {
+        if (g_scan_in_progress.exchange(true)) {
             return false;
         }
-        g_scan_in_progress = true;
         m_could_reserve = true;
         return true;
     }
```
c3cf2f55013c4ea1c1ef4a878fc7ff8e92f2c42d
It seems this changes behavior as `m_could_reserve = true;` is also guarded by mutex currently.

ajtowns replied:
`m_could_reserve` isn't a global, and instances of `CoinsViewScanReserver` aren't shared across threads either.

hebasto replied:
Oh, I see that all instances of `CoinsViewScanReserver` are local in the `scantxoutset()` function.

hebasto commented on src/wallet/wallet.h:
```diff
@@ -1286,8 +1285,7 @@ class WalletRescanReserver
     bool reserve()
     {
         assert(!m_could_reserve);
-        std::lock_guard<std::mutex> lock(m_wallet.mutexScanning);
```
960d64d326c501dbe22b80094171d7fe3e86bd69
It seems this changes behavior as the whole block is guarded by mutex currently.

ajtowns replied:
Same reasoning as previous; except that this one sets `fScanningWallet` to `true` twice :(

hebasto replied:
So why not dropping the following `m_wallet.fScanningWallet = true;` ?

hebasto replied:
As `WalletRescanReserver.m_wallet` is a reference, the object being referenced to could be shared among different threads. So not using the `mutexScanning` mutex here is not thread-safe, IMO.

ajtowns replied:
dropped redundant `fScanningWallet = true`

hebasto commented on src/net.cpp:
```diff
@@ -2366,7 +2366,7 @@ static CNetCleanup instance_of_cnetcleanup;
 {
     {
-        std::lock_guard<std::mutex> lock(mutexMsgProc);
+        LOCK(mutexMsgProc);
```
3af084a6ee2e52c660e92387bf97b024cf480c1a:
Why locking the mutex is needed here?

hebasto commented on src/test/util_threadnames_tests.cpp:
```diff
@@ -26,15 +26,17 @@ const std::string TEST_THREAD_NAME_BASE = "test_thread.";
-    std::mutex lock;
 
     auto RenameThisThread = [&](int i) {
         util::ThreadRename(TEST_THREAD_NAME_BASE + ToString(i));
-        std::lock_guard<std::mutex> guard(lock);
+        LOCK(lock);
         names.insert(util::ThreadGetInternalName());
     };
```
bbe87c91a8594e33542bae3d738499324cd06906

It seems nothing to capture now:
```suggestion
    auto RenameThisThread = [](int i) {
        util::ThreadRename(TEST_THREAD_NAME_BASE + ToString(i));
        std::lock_guard<std::mutex> guard(lock);
        LOCK(lock);
        names.insert(util::ThreadGetInternalName());
    };
```

ajtowns replied:
Dropped redundant capture

hebasto reviewed (approved):
ACK a05b44ba00b6b2328967d0ea85576abd6be381b7, tested on Linux Mint 19.3 and Clang 9.0 with `CPPFLAGS='-DDEBUG_LOCKCONTENTION -DDEBUG_LOCKORDER'`

There is the only [non-blocking nit](https://github.com/bitcoin/bitcoin/pull/16127#discussion_r426142357).

hebasto reviewed (approved):
re-ACK d61883d20a35a0de32c5825207d8b510a779b200, the only changes since the previous review are addressed nits:
- dropped redundant capture in `RenameThisThread` lambda
- dropped redundant `fScanningWallet = true;` assignment

ryanofsky commented on src/test/util_threadnames_tests.cpp:
```diff
@@ -26,15 +26,17 @@ const std::string TEST_THREAD_NAME_BASE = "test_thread.";
  */
+
+static Mutex lock;
+static std::set<std::string> names GUARDED_BY(lock);
```
In commit "test/util_threadnames_tests: add thread safety annotations" (d53072ec730d8eec5a5b72f7e65a54b141e62b19)

Assuming the reason these need to be global is lock annotations not working on local variables (didn't check), better IMO to simply not to use lock annotations here than to introduce global variables and make it unsafe to call RenameEnMasse from different threads itself. 

Other option would be to make RenameEnMasse a member function in a class with Mutex and std::set members so it could work non-globally and also be annotated

ajtowns replied:
It's not unsafe to call RenameEnMasse from different threads with the global lock, just inefficient. I tried the class approach but it seemed like overkill, so I've dropped the commit instead.

ryanofsky reviewed (approved):
Code review ACK d61883d20a35a0de32c5825207d8b510a779b200. Would call MutexGuard class LockGuard to make it an obvious replacement for [`std::lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard) and use standard names for things where Mutex is the thing being locked, and Lock is the object locking and unlocking it

ajtowns commented:
Renamed MutexGuard to LockGuard, and dropped util_threadnames_tests changes.

hebasto reviewed (approved):
re-ACK 5478d6c099e76fe070703cc5383cba7b91468b0f, only renamed s/`MutexGuard`/`LockGuard`/, and dropped the commit "test/util_threadnames_tests: add thread safety annotations" since the [previous](https://github.com/bitcoin/bitcoin/pull/16127#pullrequestreview-414184113) review.

ryanofsky reviewed (approved):
Code review ACK 5478d6c099e76fe070703cc5383cba7b91468b0f. Thanks for taking suggestions! Only changes since last review are dropping thread rename test commit d53072ec730d8eec5a5b72f7e65a54b141e62b19 and renaming mutex guard to lock guard

MarcoFalke commented:
ACK 5478d6c099e76fe070703cc5383cba7b91468b0f 🗾

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 5478d6c099e76fe070703cc5383cba7b91468b0f 🗾
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiM+Qv/bctxRqxjmwEq2xf4FSS7ej1ao6ni8G8tnk9+sB1QniYItq56EOxEe99u
lEB6lt7QFL2z9fD/mH6gEp5mdM0yqt16z35+MGxxVYNrBX0Lm/8lquFqHlPA5DNF
mfUp7Ro82ixyc65ZlMakjkgY5lr1rrNDPB9XWpw5jKFTvc5b2Dv+n7FxshM77BXd
1U8tgAOviw8etuhIWDSoTSVj85ubx9ld0cualiKg55prHi1gyH7zEQrOTJ4TYiS7
R8vPpnvK/nKpnbUq/jbXXT9DgV+3O0nKfSA+uipzsG0417ON83ugewko8oOQaP19
NtJWThsr1Eijk/YiPpdSpdNX2s0xqcvRTGaxEFU3Vwh80x4p3If2pPI96izs5Ye5
HyEGjZDtZRfco27eUp+YHPM8pk8Fg5pAJqdTQLfkGR3FTgR+r6PuXJ2GkSkXqHKv
Plx/qOYuSNlDJMd1dyCCeW6GiEpNh2XuszgldLIB8pjSCaq8qjSJtep+wmEIdsh/
ueOKDaHz
=Ww4X
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `0b593eb159203e57d289eb41d38e7938cdd641ec1ddca7c9967f9f58ff7986de  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401080b593eb159203e57d289eb41d38e7938cdd641ec1ddca7c9967f9f58ff7986def010076e49e0fdf64c9f08efab3cdcd17dbf08fff01000729c2b396617f47e7ffab0f759f0b408f1045ecef7c4f00835f85e03c086bdb00083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b1eed4f26455c9149fcbd475c371a6c208f1045ecef7c5f008368d67ad1ecc33920083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010cb1b2c27980fb97d708a17b0c088f61008f1045ecef7c6f008134680dd0ff538420083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f96397bca3c1a55b66466781671a737f08f1045ecef7c5f0086a8e138f330ee0940083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
